src/agent/advertising_proxy.cpp

```diff
@@ -0,0 +1,322 @@
+/*
+ *    Copyright (c) 2020, The OpenThread Authors.
+ *    All rights reserved.
+ *
+ *    Redistribution and use in source and binary forms, with or without
+ *    modification, are permitted provided that the following conditions are met:
+ *    1. Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *    2. Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *    3. Neither the name of the copyright holder nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ *    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *    AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *    ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *    LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *    POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   The file implements the Advertising Proxy.
+ */
+
+#include "agent/advertising_proxy.hpp"
+
+#include <string>
+
+#include <assert.h>
+
+#include "common/code_utils.hpp"
+#include "common/logging.hpp"
+
+#if OTBR_ENABLE_ADVERTISING_PROXY
+
+namespace otbr {
+
+// TODO: better to have SRP server provide separated service name labels.
+static otbrError SplitFullServiceName(const std::string &aFullName,
+                                      std::string &      aInstanceName,
+                                      std::string &      aType,
+                                      std::string &      aDomain)
+{
+    otbrError error = OTBR_ERROR_INVALID_ARGS;
+    size_t    dotPos[3];
+
+    dotPos[0] = aFullName.find_first_of('.');
```

Comment:
I think we should later but fine for me

Some ways to do it:
- Pass around `instance label` seperate from rest of service name (I like this model myself and we use it in SRP/DNS clinets)
- Escape the instance label (add `\.` instead of `.` in the instance name)  - I think this done by other implementation and suggested by the DNS-SD RFC (so we may need it if we interface with other implementations)
- Parse labels semantically (backwards) to determine the instance name (like the imp in DNS-SD server in OT).

Reply:
@simonlingoogle @abtink Thanks for sharing your thoughts. Yes I think we have chances of handling the dot in instance label by various ways. I am planning to have the SRP server provide separate instance label, service name label and domain to the advertising proxy and DNS-SD server: they share similar full service name splitting methods. I added a TODO for this method to leave it to future PRs.